test/org/apache/catalina/realm/TestRealmBase.java

```diff
@@ -789,4 +791,45 @@ public void testHttpConstraint() throws IOException {
         Assert.assertFalse(mapRealm.hasResourcePermission(
                 request, response, constraintsDelete, null));
     }
+
+    @Test
+    public void testRoleMapping() throws Exception {
+        Context context = new TesterContext() {
+            private Map<String, String> roleMapping = new HashMap<>();
+
+            public void addRoleMapping(String role, String link) {
+                roleMapping.put(role, link);
+            }
+
+            @Override
+            public String findRoleMapping(String role) {
+                return roleMapping.get(role);
+            }
+        };
+
+        context.addRoleMapping(ROLE2, "very-complex-role-name");
```

Comment:
The only way to add these Context mappings (currently) is via code. There is no mechanism to do this via configuration. Possibly as a separate enhancement, consider adding such an option. See https://bz.apache.org/bugzilla/show_bug.cgi?id=55477

Reply:
I use my fabulous [`PropertiesRoleMappingListener`](http://mo-tomcat-ext.sourceforge.net/xref/net/sf/michaelo/tomcat/extras/listeners/PropertiesRoleMappingListener.html#PropertiesRoleMappingListener) for this. I absolutely don't mind to amalgamate it into upstream somehow.

Reply:
I was thinking more along the lines of a nested element in the Context as they would be consistent with other Tomcat configuration.

Reply:
That would OK, as long as it is an interface. I don't want to hardcore those into my `context.xml`, but have them in a properties file. It could also be potentionally a database or something else.

Reply:
You probably want a `<RoleMapper class=".." />` with custom attributes. What makes it different to a listener?

Reply:
I was thinking more:
`<RoleMapping applicationName="..." realmName="..." />`
but that is a discussion better suited to BZ 55477

Reply:
I consider this way too inflexible, but yes -- let's discussed as soon as this one has been resolved.

Reply:
I will do the changes tomorrow and see how I can properly modifiy tests.